src/lib/capacityEngine.js

```diff
@@ -7,6 +7,33 @@ const { selectAsset, selectProduct, selectProductPools } = require('../store/sel
 
 const { WeiPerEther, Zero } = ethers.constants;
 
+const SECONDS_PER_DAY = BigNumber.from(24 * 3600);
+
+/**
+ * Calculates the utilization rate of the capacity.
+ *
+ * @param {Array<Object>} capacityInAssets - Array of asset objects containing assetId and amount.
+ * @param {BigNumber} capacityUsedNXM - The amount of capacity used in NXM.
+ * @returns {BigNumber} The utilization rate as a BigNumber. Returns undefined is capacity in NXM is missing
+ */
+function getUtilizationRate(capacityInAssets, capacityUsedNXM) {
+  const availableCapacityInNxm = capacityInAssets.find(asset => asset.assetId === 255)?.amount;
+  if (!availableCapacityInNxm || !capacityUsedNXM) {
+    return undefined;
+  }
+  const totalCapacity = availableCapacityInNxm.add(capacityUsedNXM);
+  return capacityUsedNXM.mul(10000).div(totalCapacity).toNumber() / 10000;
```

Comment:
jsdoc specifies this is a BigNumber, but you're returning a Number between 0 and 1 (eventually even more than 1 due to burns). I would suggest removing the `.toNumber() / 10000` part and returning that. Frontend will be able to convert that to percentage with 2 decimals.

Reply:
Secondly, did you test the edge case when `totalCapacity` is zero?

Reply:
* changed `utilizationRate` to be basis points (0-10,000)
* good catch on `totalCapacity` = 0, I'm not returning undefined.
  * it feels undefined is more appropriate in this edge case as 0 can be misinterpreted of having capacity but none is used